nlp_primitives/openai/embeddings.py

```diff
@@ -0,0 +1,186 @@
+        if max_tokens_per_batch is None:
+            self.max_tokens_per_batch = model.max_tokens * 10
+        else:
+            self.max_tokens_per_batch = max_tokens_per_batch
+
+    def _is_too_many_tokens(self, element, encoding) -> bool:
+        """Return whether a data element has too many tokens and should be ignored"""
+        return len(encoding.encode(element)) > self.model.max_tokens
+
+    def _create_request_batches(
+        self, series
+    ) -> List[OpenAIRequest[OpenAIEmbeddingResponse]]:
+        """Group elements of a series into batches of requests to send to OpenAI"""
+        # the encoding used by the model
+        encoding = tiktoken.get_encoding(self.model.encoding)
+
+        # a static embeddings response for an invalid element
+        invalid_request = StaticOpenAIEmbeddingRequest(
+            result=[np.nan] * self.number_output_features
+        )
+
+        # mutable variables to track the request batching process
+        # a running list of the requests to make
+        requests: List[OpenAIRequest[OpenAIEmbeddingResponse]] = []
+        # a list of elements that should be batched into the next request
+        elements_in_batch = []
+        # a running total of tokens that will be sent in the next request
+        tokens_in_batch = 0
+
+        def add_batched_request() -> int:
+            """Create a batched request from the currently staged elements and add it to the request list
+            """
+            if elements_in_batch:
+                elements_copy = list(elements_in_batch)
+                elements_in_batch.clear()
+                requests.append(
+                    OpenAIEmbeddingRequest(
+                        list_of_text=elements_copy,
+                        model=self.model,
+                        token_consumption=tokens_in_batch,
+                    )
+                )
+                return 0
+            else:
+                return tokens_in_batch
+
+        def can_fit_in_batch(tokens) -> bool:
+            return (
+                len(elements_in_batch) < 2048
+                and tokens_in_batch + tokens <= self.max_tokens_per_batch
+            )
+
+        # loop through the input data series to create request batches
+        for element in series:
+            if pd.isnull(element) or self._is_too_many_tokens(element, encoding):
+                # invalid element
+                # create a request from any pending elements
+                tokens_in_batch = add_batched_request()
+                # add a static request that returns the invalid results
+                requests.append(invalid_request)
+            else:
+                # valid element
+                # check how many tokens are in it
+                next_tokens = len(encoding.encode(element))
+
+                # can this element fit in the batch?
+                if can_fit_in_batch(next_tokens):
+                    # can't fit -- construct a request with existing elements
```

Comment:
Not sure if I am misunderstanding this, but should this block be the case where it can fit?

Reply:
Good call. Fixed. I need to add tests for all of this to catch stuff like this 😅